src/spoke/Spoke.sol

```diff
@@ -108,6 +102,8 @@ contract Spoke is Auth, Recoverable, ReentrancyProtection, ISpoke, ISpokeGateway
         share.authTransferFrom(msg.sender, msg.sender, address(this), amount);
         share.burn(address(this), amount);
 
+        gateway.depositSubsidy{value: msg.value}(poolId);
```

Comment:
Could we actually return the unused cost here, at this layer?

Reply:
I have a fresh idea with best of both approaches, will upload a commit with this

Reply:
Thanks for taking me along in your live thought process :D

Reply:
Before submitting, I think the most correct solution is:

```solidity
uint256 cost = sender.sendRegisterAsset(centrifugeId, assetId, decimals);
        if (cost == 0) {
            payable(msg.sender).call{value: msg.value}("");
        } else {
            gateway.depositSubsidy{value: cost}(PoolId.wrap(0));
        }
```

- If the user doesn't pay `cost` and the message was sent, it will fail.
- If the user pays `cost` but not subsidized funds, the message is added to the unpaid, and the user will be refunded, to later allow them to call `repay()` without losing funds.
- The pool is subsidized with `cost` if the message was sent.


The other option is to fail if no subsidized funds. Tell me if you agree.

Reply:
Although, in most cases, the user will call `crosschainTransferShares{value: exactCost}()` which makes me lean towards solution 2. again because it will never fail due a unsubsidized pool:

```solidity
gateway.depositSubsidy{value: cost}(PoolId.wrap(0));
uint256 cost = sender.sendRegisterAsset(centrifugeId, assetId, decimals);
require(msg.value > cost);
```

Reply:
Sorry for doing so many back and forth with this. After testing almost every possible solution. I decay for option 2 like:

```
gateway.depositedFunds{value: msg.value}()
uint256 cost = sender.send*();
require(msg.value >= cost)
```

- The thing is that we never want the state where some integrator call fails because lack of subsidized funds if they're already providing those.
- Also, the caller needs a way to estimate the call to at least `cost` value.

The point is that using `cost` is to save us from estimate inside the method, but it will not save the caller to anyways estimate if the need to pass some gas.

Reply:
It would be cool to provide a mechanism to refund any value > cost from the gateway from the caller. Something like:

```
gateway.withdrawFunds(msg.value - cost, msg.sender)
```

As long as the integrator is also a gateway manager, they can do that.

Reply:
Agree with going to option 2!